// @galacean/engine-toolkit-gizmo@1.4.0-alpha.1/dist/es/Type.js
import { _ as _inherits } from './_inherits-5a51c3d0.js';
import { Vector3, Plane, Component } from '@galacean/engine';

/**
 * @internal
 * Gizmo Component
 */ var GizmoComponent = /*#__PURE__*/ function(Component) {
    _inherits(GizmoComponent, Component);
    function GizmoComponent() {
        return Component.apply(this, arguments);
    }
    return GizmoComponent;
}(Component);
var axisType;
(function(axisType) {
    axisType[axisType["x"] = 0] = "x";
    axisType[axisType["y"] = 1] = "y";
    axisType[axisType["z"] = 2] = "z";
    axisType[axisType["xyz"] = 3] = "xyz";
    axisType[axisType["xy"] = 4] = "xy";
    axisType[axisType["yz"] = 5] = "yz";
    axisType[axisType["xz"] = 6] = "xz";
})(axisType || (axisType = {}));
var axisVector = [
    new Vector3(1, 0, 0),
    new Vector3(0, 1, 0),
    new Vector3(0, 0, 1),
    new Vector3(1, 1, 1),
    new Vector3(1, 1, 0),
    new Vector3(0, 1, 1),
    new Vector3(1, 0, 1)
];
var axisPlane = [
    new Plane(new Vector3(1, 0, 0), 0),
    new Plane(new Vector3(0, 1, 0), 0),
    new Plane(new Vector3(0, 0, 1), 0),
    new Plane(new Vector3(0, 0, 0), 0),
    new Plane(new Vector3(0, 0, 1), 0),
    new Plane(new Vector3(1, 0, 0), 0),
    new Plane(new Vector3(0, 1, 0), 0)
];

export { GizmoComponent, axisPlane, axisType, axisVector };
