// @galacean/engine-toolkit-stats@1.4.0-alpha.1/dist/es/hooks/TextureHook.js
import { log } from '../log.js';
import '@galacean/engine';

var TextureHook = /*#__PURE__*/ function() {
    function TextureHook(gl) {
        this.textures = 0;
        this.realCreateTexture = gl.createTexture;
        this.realDeleteTexture = gl.deleteTexture;
        gl.createTexture = this.hookedCreateTexture.bind(this);
        gl.deleteTexture = this.hookedDeleteTexture.bind(this);
        this.hooked = true;
        this.gl = gl;
        log("Texture is hooked.");
    }
    var _proto = TextureHook.prototype;
    _proto.hookedCreateTexture = function hookedCreateTexture() {
        var texture = this.realCreateTexture.call(this.gl);
        this.textures++;
        log("CreateTexture:", texture, "textures: " + this.textures);
        return texture;
    };
    _proto.hookedDeleteTexture = function hookedDeleteTexture(texture) {
        this.realDeleteTexture.call(this.gl, texture);
        this.textures--;
        log("DeleteTexture. textures: " + this.textures);
    };
    _proto.reset = function reset() {
        this.textures = 0;
    };
    _proto.release = function release() {
        if (this.hooked) {
            this.gl.createTexture = this.realCreateTexture;
            this.gl.deleteTexture = this.realDeleteTexture;
        }
        this.hooked = false;
    };
    return TextureHook;
}();

export { TextureHook as default };
